volume/store/store.go

```diff
@@ -536,16 +545,18 @@ func (s *VolumeStore) Dereference(v volume.Volume, ref string) {
 	s.locks.Lock(v.Name())
 	defer s.locks.Unlock(v.Name())
 
-	s.globalLock.Lock()
-	defer s.globalLock.Unlock()
 	var refs []string
-
+	s.globalLock.RLock()
 	for _, r := range s.refs[v.Name()] {
 		if r != ref {
 			refs = append(refs, r)
 		}
 	}
+	s.globalLock.RUnlock()
+
+	s.globalLock.Lock()
 	s.refs[v.Name()] = refs
+	s.globalLock.Unlock()
```

Comment:
yes, you are right, this may cause some concurrent delete reference actions to be racing, and som refs were not deleted, thx for comment.